hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -844,43 +840,79 @@ class Emit[C](
+          cb += eltRegion.allocateRegion(Region.REGULAR)
+          cb += tmpRegion.allocateRegion(Region.REGULAR)
+          cb.assign(xAcc, emitI(zero, eltRegion).map(cb)(_.castTo(mb, eltRegion.code, accType)))
+
+          stream.asStream.stream.getStream(eltRegion).forEachI(cb, { elt =>
+            // pre- and post-condition: 'xAcc' contains current accumulator,
+            // whose heap memory is contained in 'eltRegion'. 'tmpRegion' is
+            // empty.
             cb.assign(xElt, elt)
-            cb.assign(xAcc, emitI(body, env = env.bind(accumName -> xAcc, valueName -> xElt))
-              .map(cb)(_.castTo(mb, region.code, accType)))
+            cb.assign(xAcc, emitI(body, eltRegion, env.bind(accumName -> xAcc, valueName -> xElt))
+              .map(cb)(eltRegion.copyToSibling(mb, _, tmpRegion, accType)))
+            cb += eltRegion.clear()
+            cb += StagedRegion.swap(mb, eltRegion, tmpRegion)
           })
 
+          cb += tmpRegion.free()
+          cb.assign(xAcc, xAcc.map(eltRegion.copyToParent(mb, _)))
+          cb += eltRegion.free()
+
           xAcc.toI(cb)
         }
 
       case x@StreamFold2(a, acc, valueName, seq, res) =>
         val eltType = coerce[PStream](a.pType).elementType
+
         val xElt = mb.newEmitField(valueName, eltType)
         val names = acc.map(_._1)
         val accTypes = x.accPTypes
         val accVars = (names, accTypes).zipped.map(mb.newEmitField)
+        val tmpAccVars = (names, accTypes).zipped.map(mb.newEmitField)
+
+        val eltRegion = region.createChildRegion(mb)
+        val tmpRegion = region.createChildRegion(mb)
 
         val resEnv = env.bind(names.zip(accVars): _*)
         val seqEnv = resEnv.bind(valueName, xElt)
 
         val streamOpt = emitStream(a)
         streamOpt.flatMap(cb) { stream =>
-          (accVars, acc, accTypes).zipped.foreach { case (xAcc, (_, x), typ) =>
-            cb.assign(xAcc, emitI(x).map(cb)(_.castTo(mb, region.code, typ)))
+          cb += eltRegion.allocateRegion(Region.REGULAR)
+          cb += tmpRegion.allocateRegion(Region.REGULAR)
+
+          (accVars, acc).zipped.foreach { case (xAcc, (_, x)) =>
+            cb.assign(xAcc, emitI(x, eltRegion).map(cb)(_.castTo(mb, eltRegion.code, xAcc.pt)))
           }
-          stream.asStream.stream.getStream(region).forEachI(cb, { elt =>
+          stream.asStream.stream.getStream(eltRegion).forEachI(cb, { elt =>
+            // pre- and post-condition: 'accVars' contain current accumulators,
+            // all of whose heap memory is contained in 'eltRegion'. 'tmpRegion'
+            // is empty.
             cb.assign(xElt, elt)
-            (accVars, seq).zipped.foreach { (accVar, ir) =>
-              cb.assign(accVar, emitI(ir, env = seqEnv).map(cb)(_.castTo(mb, region.code, accVar.pt)))
+            (tmpAccVars, seq).zipped.foreach { (accVar, ir) =>
+              // FIXME: needs to be simultaneous assignment
```

Comment:
what does this mean?

Reply:
Whoops, that's old. The two stage assignment to `tmpAccVars` then to `accVars` addresses this, so all `seq` IRs are evaluated in an environment with the previous accumulator values. I'll delete.

Reply:
ok, cool. Looked right to me.